### PR TITLE
Allow parameter overrides for a DB cluster parameter group

### PR DESCRIPTION
**Pull request by jljaco**
Issue #, if available:
https://github.com/aws-controllers-k8s/community/issues/869

The `Spec.ParameterOverrides` is a `map[string]*string` containing the parameter names and values for parameters in a DB cluster parameter group for which the user wishes to override the defaults.

The `Status.ParameterOverrideStatuses` field is a slice of `Parameter` structs that contains status information about those overridden parameters.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jljaco** [comment]
This is still a work in progress

File: config/crd/bases/rds.services.k8s.aws_dbclusterparametergroups.yaml [hunk 1]
```diff
@@ -192,6 +153,41 @@ spec:
                   - type
                   type: object
                 type: array
+              parameterOverrideStatuses:
+                description: Provides a list of parameters for the DB cluster parameter
+                  group.
+                items:
+                  description: "This data type is used as a request parameter in the
```
Comment (jljaco) [inline config/crd/bases/rds.services.k8s.aws_dbclusterparametergroups.yaml:160]: This misleading comment extracted automatically from https://github.com/aws/aws-sdk-go/blob/6cdd6178046217238d1969f5343fd929d7173703/service/rds/api.go#L44729

File: helm/crds/rds.services.k8s.aws_dbclusterparametergroups.yaml [hunk 2]
```diff
@@ -193,6 +154,41 @@ spec:
                   - type
                   type: object
                 type: array
+              parameterOverrideStatuses:
+                description: Provides a list of parameters for the DB cluster parameter
+                  group.
+                items:
+                  description: "This data type is used as a request parameter in the
```
Comment (jljaco) [inline helm/crds/rds.services.k8s.aws_dbclusterparametergroups.yaml:161]: Ditto -- automatically extracted from https://github.com/aws/aws-sdk-go/blob/6cdd6178046217238d1969f5343fd929d7173703/service/rds/api.go#L44729

File: pkg/resource/db_cluster_parameter_group/delta.go [hunk 3]
```diff
@@ -63,8 +63,12 @@ func newResourceDelta(
 			delta.Add("Spec.Name", a.ko.Spec.Name, b.ko.Spec.Name)
 		}
 	}
-	if !reflect.DeepEqual(a.ko.Spec.Parameters, b.ko.Spec.Parameters) {
-		delta.Add("Spec.Parameters", a.ko.Spec.Parameters, b.ko.Spec.Parameters)
+	if ackcompare.HasNilDifference(a.ko.Spec.ParameterOverrides, b.ko.Spec.ParameterOverrides) {
```
Comment (jljaco) [inline pkg/resource/db_cluster_parameter_group/delta.go:66]: Noting this comment from the analogous line in #107:

https://github.com/aws-controllers-k8s/rds-controller/pull/107/files#r948248230

**Comment by jljaco** [comment]
Noting that the e2e tests are currently [failing](https://prow.ack.aws.dev/view/s3/ack-prow-logs/pr-logs/pull/aws-controllers-k8s_rds-controller/133/rds-kind-e2e/1633544817986244608) for an unrelated reason.

The assertion [here](https://github.com/aws-controllers-k8s/rds-controller/blob/c288c4f1e60fd5d9c86ef294690bd87da4bf007c/test/e2e/tests/test_db_proxy.py#L79) is failing.

I am specifically not going to fix that problem in this PR; it should be fixed in another one.

File: apis/v1alpha1/db_cluster_parameter_group.go [hunk 4]
```diff
@@ -88,20 +88,8 @@ type DBClusterParameterGroupSpec struct {
...
-	// the pending-reboot value for both dynamic and static parameters.
-	//
-	// When the application method is immediate, changes to dynamic parameters are
-	// applied immediately to the DB clusters associated with the parameter group.
-	// When the application method is pending-reboot, changes to dynamic and static
-	// parameters are applied after a reboot without failover to the DB clusters
-	// associated with the parameter group.
-	Parameters []*Parameter `json:"parameters,omitempty"`
+	Name               *string            `json:"name"`
+	ParameterOverrides map[string]*string `json:"parameterOverrides,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/db_cluster_parameter_group.go:92]: We 100% cannot remove, or modify the shape of, an existing spec field. It breaks backwards compatibility and would have pretty strange effects for existing customers. I understand this complicates the implementation, but if we can fit the changes within the existing spec that would be nice, otherwise we have to add a new field altogether.

File: pkg/resource/db_cluster_parameter_group/hooks.go [hunk 5]
```diff
@@ -173,3 +196,234 @@ func sdkTagsFromResourceTags(
...
+	err error,
+) {
+	var marker *string
+	params = make(map[string]*string)
+	for {
+		resp, err := rm.sdkapi.DescribeDBClusterParametersWithContext(
+			ctx,
+			&svcsdk.DescribeDBClusterParametersInput{
+				DBClusterParameterGroupName: groupName,
+				Source:                      aws.String(sourceUser),
```
Comment (RedbackThomson) [inline pkg/resource/db_cluster_parameter_group/hooks.go:261]: I have read the RDS docs but I still don't understand the significance of this. Is this relevant to our controller at all?

File: pkg/util/parameters.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	// from their list of parameter overrides, we will not be able to get the
+	// resource into a synced state.
+	return ackerr.NewTerminalError(
+		fmt.Errorf("%w: %s", errUnmodifiableParameter, name),
+	)
+}
+
+// ComputeParametersDelta compares two Parameter arrays and returns the new
+// parameters to add, to update and the parameter identifiers to delete
+func ComputeParametersDelta(
```
Comment (RedbackThomson) [inline pkg/util/parameters.go:36]: Check out how I did this for tags in the runtime using `lo` helper methods - https://github.com/aws-controllers-k8s/runtime/blob/main/pkg/compare/tags.go#L26

It might not be any shorter, but I think the code is more maintainable

File: pkg/util/parameters.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+		if _, found := desired[k]; !found {
+			toReset = append(toReset, k)
+		}
+	}
+	return toModify, toReset
+}
+
+// sliceStringChunks splits a supplied slice of string pointers into multiple
+// slices of string pointers of a given size.
+func SliceStringChunks(
```
Comment (RedbackThomson) [inline pkg/util/parameters.go:60]: `samber/lo` supports this out of the box - https://github.com/samber/lo#chunk

File: pkg/util/parameters.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+			chunk = map[string]*string{}
+			idx = 0
+		}
+	}
+	chunks = append(chunks, chunk)
+
+	return chunks
+}
+
+// *** Caching
```
Comment (RedbackThomson) [inline pkg/util/parameters.go:106]: nit: move all of this under a new file `caching.go`

File: pkg/util/parameters.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
+package util
```
Comment (RedbackThomson) [inline pkg/util/parameters.go:1]: This file needs an Amazon license header

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
Really excellent and clean code with the custom hook refactoring. Unfortunately there is a major issue with modifying the existing API shapes and/or names, that I've commented on below.

File: apis/v1alpha1/db_cluster_parameter_group.go (hunk 4 above)
Comment (jljaco) [reply apis/v1alpha1/db_cluster_parameter_group.go:92]: OK, just to be clear, is this only the rule for GA controllers?

Because an analogous change was made last September for `db_parameter_group` [here](https://github.com/aws-controllers-k8s/rds-controller/pull/107/files#diff-fbd5c921ed4d5f67af22264e781b19e636ba01ca7e8c8c95c043bfb8ccb3ecb2)

Comment (A-Hilaly) [reply apis/v1alpha1/db_cluster_parameter_group.go:92]: >OK, just to be clear, is this only the rule for GA controllers?

Yes.

>Because an analogous change was made last September for db_parameter_group [here](https://github.com/aws-controllers-k8s/rds-controller/pull/107/files#diff-fbd5c921ed4d5f67af22264e781b19e636ba01ca7e8c8c95c043bfb8ccb3ecb2)

That change was made few days before going GA :/. If we don't have multi-version support we cannot delete/rename any fields.

File: pkg/util/parameters.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+	"context"
+	"fmt"
+	"sync"
+
+	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+)
+
+var (
+	errUnknownParameter      = fmt.Errorf("unknown parameter")
+	errUnmodifiableParameter = fmt.Errorf("parameter is not modifiable")
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:13]: nit: i know those errors are not imported anywhere, but since they are pre-defined errors they need to be exposed (public) in case in one wants to select types on them.

File: pkg/util/parameters.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+func NewErrUnknownParameter(name string) error {
+	// This is a terminal error because unless the user removes this parameter
+	// from their list of parameter overrides, we will not be able to get the
+	// resource into a synced state.
+	return ackerr.NewTerminalError(
+		fmt.Errorf("%w: %s", errUnknownParameter, name),
+	)
+}
+
+func NewErrUnmodifiableParameter(name string) error {
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:25]: Public functions/objects needs to have a comment - https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

File: pkg/util/parameters.go [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	for k := range latest {
+		if _, found := desired[k]; !found {
+			toReset = append(toReset, k)
+		}
+	}
+	return toModify, toReset
+}
+
+// sliceStringChunks splits a supplied slice of string pointers into multiple
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:58]:
```suggestion
// SliceStringChunks splits a supplied slice of string pointers into multiple
```

File: pkg/util/parameters.go [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+		}
+
+		chunks = append(chunks, input[0:chunkSize])
+		input = input[chunkSize:]
+	}
+
+	return chunks
+}
+
+// mapStringChunks splits a supplied map of string pointers into multiple
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:81]:
```suggestion
// MapStringChunks splits a supplied map of string pointers into multiple
```

File: pkg/util/parameters.go [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+		}
+	}
+	chunks = append(chunks, chunk)
+
+	return chunks
+}
+
+// *** Caching
+
+type ParamMeta struct {
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:108]: ditto

File: pkg/util/parameters.go [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+	IsModifiable bool
+	IsDynamic    bool
+}
+
+// metaFetcher is the functor we pass to the paramMetaCache that allows it to
+// fetch engine default parameter information
+type metaFetcher func(ctx context.Context, family string) (map[string]ParamMeta, error)
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:115]: Preferably `MetaFetcher` needs to be exported

**Review by A-Hilaly — COMMENTED** [review]
Good stuff! left few tiny suggestions.

File: pkg/resource/db_cluster_parameter_group/hooks.go (hunk 5 above)
Comment (jaypipes) [reply pkg/resource/db_cluster_parameter_group/hooks.go:261]: @RedbackThomson there's no need to return any non-user-settable parameters, which is why this filter is passed.

File: pkg/util/parameters.go (hunk 8 above)
Comment (jaypipes) [reply pkg/util/parameters.go:106]: Or `param_cache.go`

**Review by jaypipes — COMMENTED** [review]
It's a shame the RDS controller was GA'd before this could be done. I almost prefer to make the backwards incompatible CRD schema changes so that this aligns with the DBParameterGroup CRD... especially since we're still at v1alpha1 API version...

File: apis/v1alpha1/db_cluster_parameter_group.go (hunk 4 above)
Comment (jljaco) [reply apis/v1alpha1/db_cluster_parameter_group.go:92]: As discussed in the community meeting this morning, it is worth noting that this is a field that hasn't actually ever worked.  I'm a bit dubious that anyone might actually be using it.  Also, there is a value to having `db_parameter_group` and `db_cluster_parameter_group` behave similarly, since they're both effectively doing the same thing.

File: pkg/util/parameters.go (hunk 9 above)
Comment (jljaco) [reply pkg/util/parameters.go:14]: Done

File: pkg/util/parameters.go (hunk 8 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

(`parameter_cache.go`)

File: pkg/util/parameters.go (hunk 10 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

File: pkg/util/parameters.go (hunk 11 above)
Comment (jljaco) [reply pkg/util/parameters.go:40]: Done

File: pkg/util/parameters.go (hunk 13 above)
Comment (jljaco) [reply pkg/util/parameters.go:96]: Done

File: pkg/util/parameters.go (hunk 15 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

File: pkg/util/parameters.go (hunk 12 above)
Comment (jljaco) [reply pkg/util/parameters.go:73]: Done

File: pkg/util/parameters.go (hunk 7 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

**Review by jaypipes — CHANGES_REQUESTED** [review]
@jljaco I believe you still need to 1) add a call to `syncParameters` somewhere (I don't see either `customUpdate` changed nor do I see any template/hook code call `syncParameters` so parameter overrides will never be set. 2) modify `test/e2e/tests/test_db_cluster_parameter_group.py` to add the parameter override testing

File: apis/v1alpha1/db_cluster_parameter_group.go (hunk 4 above)
Comment (jljaco) [reply apis/v1alpha1/db_cluster_parameter_group.go:93]: Restored `Parameters` in `generator.yaml` and marked it as `DEPRECATED` using custom doc strings

File: go.local.mod [hunk 16]
```diff
@@ -50,10 +50,10 @@ require (
...
 	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
 	github.com/pkg/errors v0.9.1 // indirect
-	github.com/prometheus/client_golang v1.14.0 // indirect
-	github.com/prometheus/client_model v0.3.0 // indirect
-	github.com/prometheus/common v0.37.0 // indirect
-	github.com/prometheus/procfs v0.8.0 // indirect
+	github.com/prometheus/client_golang v1.11.0 // indirect
+	github.com/prometheus/client_model v0.2.0 // indirect
+	github.com/prometheus/common v0.28.0 // indirect
+	github.com/prometheus/procfs v0.6.0 // indirect
```
Comment (A-Hilaly) [inline go.local.mod:56]: Maybe a rebase issue? client_golang should stay `v1.14.0`. Source: https://github.com/aws-controllers-k8s/runtime/blob/main/go.mod#L13

Comment (A-Hilaly) [reply go.local.mod:56]: Mmmmm looks like this the local one... please discard my comment :)

File: pkg/util/parameters.go (hunk 6 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

File: templates/hooks/db_cluster_parameter_group/sdk_read_many_post_set_output.go.tpl [hunk 17]
```diff
@@ -1,8 +1,17 @@
...
+    }
+    if ko.Spec.Name != nil {
+        groupName := ko.Spec.Name
+        params, paramStatuses, err := rm.getParameters(ctx, groupName)
+        if err != nil {
+            return nil, err
+        }
+        ko.Spec.ParameterOverrides = params
+        ko.Status.ParameterOverrideStatuses = paramStatuses
+    }
```
Comment (jaypipes) [inline templates/hooks/db_cluster_parameter_group/sdk_read_many_post_set_output.go.tpl:17]: Not necessary to fix in this PR, but just a quick tip... you should replace four spaces with `\t` characters in all the templates. Likely your IDE or text editor just isn't using TABs when the code file isn't a `.go` file...

**Review by jaypipes — APPROVED** [review]
I think this is ready to go. Thank you @jljaco :)

File: pkg/util/parameters.go [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+var (
+	ErrUnknownParameter      = fmt.Errorf("unknown parameter")
+	ErrUnmodifiableParameter = fmt.Errorf("parameter is not modifiable")
+)
+
+// Parameters represents the elements of a DB Parameter Group
+// or a DB Cluster Parameter Group
+type Parameters map[string]*string
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:30]: Type aliasing for basic Go type is very rarely recommended in idiomatic Go - i'd suggest just use `map[string]*string` in the function parameters

File: pkg/util/parameters.go [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+	return added, unchanged, removed
+}
+
+// ChunkMap splits a supplied map of parameters into multiple
+// slices of maps of parameters of a given size.
+func ChunkMap(
+	input Parameters,
+	chunkSize int,
+) []Parameters {
```
Comment (A-Hilaly) [inline pkg/util/parameters.go:79]: `ChunkMap` is good name if input was `map[string]*string`, otherwise it should be `ChunkParameters`

File: pkg/resource/db_cluster_parameter_group/hooks.go [hunk 20]
```diff
@@ -15,21 +15,44 @@ package db_cluster_parameter_group
...
 
 	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
 	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
 
 	svcapitypes "github.com/aws-controllers-k8s/rds-controller/apis/v1alpha1"
+	"github.com/aws/aws-sdk-go/aws"
 	svcsdk "github.com/aws/aws-sdk-go/service/rds"
 
 	"github.com/aws-controllers-k8s/rds-controller/pkg/util"
```
Comment (A-Hilaly) [inline pkg/resource/db_cluster_parameter_group/hooks.go:29]: https://github.com/golang/go/wiki/CodeReviewComments#imports

File: pkg/util/parameter_cache.go [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+// family. We use this cached information to determine whether a parameter is
+// statically or dynamically defined (whether changes can be applied
+// immediately or pending a reboot) and whether a parameter is modifiable.
+//
+// Keeping things super simple for now and not adding any TTL or expiration
+// behaviour to the cache. Engine defaults are pretty static information...
+type ParamMetaCache struct {
+	sync.RWMutex
+	Hits   uint64
+	Misses uint64
```
Comment (A-Hilaly) [inline pkg/util/parameter_cache.go:41]: @jaypipes I still don't understand the use of Hits and Missed in this implementation, they are nor used nor logged, just incremented - did you have other further intentions for this?

File: test/e2e/tests/test_db_cluster_parameter_group.py [hunk 22]
```diff
@@ -82,8 +83,14 @@ def test_create_delete_aurora_mysql5_7(self):
...
         ]
+        new_params = {
+            "autocommit": "1",
+        }
         updates = {
-            "spec": {"tags": new_tags},
+            "spec": {
+                "tags": new_tags,
+                "parameterOverrides": new_params,
+            },
```
Comment (A-Hilaly) [inline test/e2e/tests/test_db_cluster_parameter_group.py:93]: Since this implementation is exceptionally complex - i'd suggest testing more than just one parameter.

**Review by A-Hilaly — CHANGES_REQUESTED** [review]
Good job on this! left few comments inline

File: pkg/util/parameters.go (hunk 18 above)
Comment (jljaco) [reply pkg/util/parameters.go:30]: https://github.com/aws-controllers-k8s/runtime/blob/fb553741e0bfbbd1fee8f52b2e36b8c58dd7bf8e/pkg/tags/tags.go#L21

My intent was to create an analog to the [GetTagsDifference](https://github.com/aws-controllers-k8s/runtime/blob/7152103f915b2722bef679ebe6df231e7f489222/pkg/compare/tags.go#L26) function

as [suggested here](https://github.com/aws-controllers-k8s/rds-controller/pull/133#discussion_r1130237940)

File: pkg/util/parameters.go (hunk 19 above)
Comment (jljaco) [reply pkg/util/parameters.go]: Done

File: pkg/resource/db_cluster_parameter_group/hooks.go [hunk 23]
```diff
@@ -173,3 +201,235 @@ func sdkTagsFromResourceTags(
 	}
 	return tags
 }
+
+func equalStrings(a, b *string) bool {
+	if a == nil {
+		return b == nil || *b == ""
+	}
+	return (*a == "" && b == nil) || *a == *b
+}
```
[the remark this answers is not in the document]
Comment (jljaco) [reply pkg/resource/db_cluster_parameter_group/hooks.go:210]: Noted.  As you've pointed out in that comment, the issue applies to many different instances of this function across the codebase.  Would you prefer that I fix this one instance, knowing that it will be out of alignment with all of the other, broken instances?

[the remark this answers is not in the document]
Comment (A-Hilaly) [reply pkg/resource/db_cluster_parameter_group/hooks.go:210]: Ideally we should fix it now. Let's not wait until it lands in runtime. Updating all the controllers will take alot of time.

[the remark this answers is not in the document]
Comment (jljaco) [reply pkg/resource/db_cluster_parameter_group/hooks.go]: Turns out that this function was unused, so I just pruned it.

File: pkg/util/parameter_cache.go [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+	name string,
+	fetcher MetaFetcher,
+) (*ParamMeta, error) {
+	var err error
+	var found bool
+	var metas map[string]ParamMeta
+	var meta ParamMeta
+
+	// We need to release the lock right after the read operation, because
+	// loadFamilly will might call a writeLock at L619
```
Comment (RedbackThomson) [inline pkg/util/parameter_cache.go:59]: nit: Update the line number in this comment to match new file location

**Review by RedbackThomson — APPROVED** [review]
Really really excellent work.
If I were to nitpick the testing, I might say that we should test adding params in the initial create - so we can test that hook code you've added for requeueing instantly after create. Also possibly adding another patch to remove a param, so we know resetting parameters works.

File: pkg/util/parameter_cache.go (hunk 24 above)
Comment (jljaco) [reply pkg/util/parameter_cache.go]: Removed the line number entirely

File: test/e2e/tests/test_db_cluster_parameter_group.py [hunk 25]
```diff
@@ -97,15 +97,18 @@ def test_create_delete_aurora_mysql5_7(
...
+                found += 1
                 assert "ParameterValue" in tp, f"No ParameterValue in parameter of name 'aurora_binlog_read_buffer_size': {tp}"
                 assert tp["ParameterValue"] == "8192", f"Wrong value for parameter of name 'aurora_binlog_read_buffer_size': {tp}"
-                break
-        assert found, f"No parameter of name 'aurora_binlog_read_buffer_size' was found: {test_params}"
+            elif tp["ParameterName"] == "aurora_read_replica_read_committed":
+                found += 1
+                assert "ParameterValue" in tp, f"No ParameterValue in parameter of name 'aurora_read_replica_read_committed': {tp}"
+                assert tp["ParameterValue"] == "OFF", f"Wrong value for parameter of name 'aurora_read_replica_read_committed': {tp}"
+        assert found == 2, f"Did not find parameters with names 'aurora_binlog_read_buffer_size' and 'aurora_read_replica_read_committed': {test_params}"
```
Comment (A-Hilaly) [inline test/e2e/tests/test_db_cluster_parameter_group.py:111]: QQ: I guess you're iterating on the array (and using `found`var) because the order of `tp` is not deterministic?

**Review by A-Hilaly — APPROVED** [review]
Looks good and clean! Thank you @jljaco

**Comment by james-callahan** [comment]
It appears this may have been reverted. The CRD comment particularly in https://github.com/aws-controllers-k8s/rds-controller/pull/155/files#diff-2b8488e2b84c0c1089daa216c5df7bd1caadee6e2af5e378590957856bae5560L107
